### ISSUE TITLE
Template Literal Type - Does not evaluate interpolated template strings against type definition

### ISSUE THREAD

**Issue by dan-kez** [microsoft/TypeScript#41732]
**TypeScript Version:**  4.1.2

**Search Terms:**
template literal type template string interpolation number pixel value 

**Code**
```ts
const pixelValue: number = 22;

type PixelValueType = `${number}px`;

const pixelString: PixelValueType = `22px`;

// Error here
const pixelStringWithTemplate: PixelValueType = `${pixelValue}px`;
```

**Expected behavior:**
A template string should be evaluated to determine if it matches a template literal type.

**Actual behavior:**
The template string is evaluated to type `string` even though it can match the Template Literal Type
```
Type 'string' is not assignable to type '`${number}px`'.(2322)
```

**Playground Link:** 
https://www.typescriptlang.org/play?#code/MYewdgzgLgBADgSwB4FMA2A1AhmgrigLhjFwFsAjFAJxgF4YAmBgbgChWoBPOFGABWTpseFABVuvegAMAJAG8SFagF84SKW1ahIsRKjQBlKFQRgA5kQH7h+cTzowpTNRq3ho8QYeOmzAdQQoAAtRFFI4NCwoQn4vGzEJB1k5PSEcfFV1NiA

**Related Issues:**

**Comment by jcalz** [comment]
Related issues: #40707, #10676

From #40707, if you use a `const` assertion the compiler will give a template literal type to a template literal expression, resulting in your desired behavior:

```ts
const pixelValue: number = 22;
type PixelValueType = `${number}px`;
const pixelStringWithTemplate: PixelValueType = `${pixelValue}px` as const; // okay
```

[Playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBADgSwB4FMA2A1AhmgrigLhjFwFsAjFAJxgF4YAmBgbgCgoBPOFGABWXTY8KACpce9AAYASAN4kK1AL5wkktqEixEqNAGUoVBGADmAdQRQAFiJSk4aLFEJ8BmHPjHc6MGbJ2CHigqajBYEDCa0MwwAPSxMCAA1lgcrKxAA)

It looks like your desired behavior is for the compiler to always synthesize a template literal type for a template literal expression the way #10676 does for other literal types, so that `as const` is not necessary.

**Comment by a1245582339** [comment]
> Related issues: #40707, #10676
> 
> From #40707, if you use a `const` assertion the compiler will give a template literal type to a template literal expression, resulting in your desired behavior:
> 
> ```ts
> const pixelValue: number = 22;
> type PixelValueType = `${number}px`;
> const pixelStringWithTemplate: PixelValueType = `${pixelValue}px` as const; // okay
> ```
> 
> [Playground link](https://www.typescriptlang.org/play?#code/MYewdgzgLgBADgSwB4FMA2A1AhmgrigLhjFwFsAjFAJxgF4YAmBgbgCgoBPOFGABWXTY8KACpce9AAYASAN4kK1AL5wkktqEixEqNAGUoVBGADmAdQRQAFiJSk4aLFEJ8BmHPjHc6MGbJ2CHigqajBYEDCa0MwwAPSxMCAA1lgcrKxAA)
> 
> It looks like your desired behavior is for the compiler to always synthesize a template literal type for a template literal expression the way #10676 does for other literal types, so that `as const` is not necessary.

Why the `${pixelValue}px` cannot be inferred to PixelValueType automatically? It seems nothing problem.

**Comment by RyanCavanaugh** [comment]
@ahejlsberg Seems like we should be providing a literal context here?

**Comment by ahejlsberg** [comment]
This is a duplicate of #41631.

**Comment by ahejlsberg** [comment]
Now implemented in #41891.
